pkg/auth/authenticator/request/websocket/protocol.go

```diff
@@ -0,0 +1,82 @@
+package websocket
+
+import (
+	"errors"
+	"net/http"
+	"net/textproto"
+	"regexp"
+	"strings"
+	"unicode/utf8"
+
+	"encoding/base64"
+
+	"github.com/openshift/origin/pkg/auth/authenticator"
+	"k8s.io/apiserver/pkg/authentication/user"
+)
+
+const bearerProtocolPrefix = "base64.bearer.authorization."
+
+var protocolHeader = textproto.CanonicalMIMEHeaderKey("Sec-WebSocket-Protocol")
+
+// ProtocolAuthenticator allows a websocket connection to provide a bearer token as a subprotocol
+// in the format "bearer.authorization.openshift.io.<base64url-encoded-token-with-no-padding>"
+type ProtocolAuthenticator struct {
+	// auth is the token authenticator to use to validate the token
+	auth authenticator.Token
+	// remove indicates whether the protocol with the token should be stripped from the incoming request
+	remove bool
+}
+
+func NewProtocolAuthenticator(auth authenticator.Token, remove bool) *ProtocolAuthenticator {
+	return &ProtocolAuthenticator{auth, remove}
+}
+
+func (a *ProtocolAuthenticator) AuthenticateRequest(req *http.Request) (user.Info, bool, error) {
+	// Only accept websocket connections
+	if !isWebSocketRequest(req) {
+		return nil, false, nil
+	}
+
+	token := ""
+	filteredProtocols := []string{}
+	for _, protocolHeader := range req.Header[protocolHeader] {
+		for _, protocol := range strings.Split(protocolHeader, ",") {
+			protocol = strings.TrimSpace(protocol)
+
+			if !strings.HasPrefix(protocol, bearerProtocolPrefix) {
+				filteredProtocols = append(filteredProtocols, protocol)
+				continue
+			}
+
+			encodedToken := strings.TrimPrefix(protocol, bearerProtocolPrefix)
+			decodedToken, err := base64.RawURLEncoding.DecodeString(encodedToken)
+			if err != nil {
+				return nil, false, errors.New("invalid token encoding")
+			}
+			if !utf8.Valid(decodedToken) {
+				return nil, false, errors.New("invalid token bytes")
+			}
+			token = string(decodedToken)
+		}
+	}
+
+	if len(token) == 0 {
+		return nil, false, nil
+	}
+
+	user, ok, err := a.auth.AuthenticateToken(token)
+	if ok && a.remove {
+		req.Header.Set(protocolHeader, strings.Join(filteredProtocols, ","))
```

Comment:
Link to the spec that indicates that repeating the header is equivalent to having the header once with a comma delimited list?